core/shared/src/main/scala-3/zio/internal/macros/LayerMacroUtils.scala

```diff
@@ -42,31 +42,27 @@ private[zio] object LayerMacroUtils {
       loop(TypeRepr.of[T])
     }
 
-    val targetTypes    = getRequirements[R]
-    val remainderTypes = getRequirements[R0]
     val layerToDebug: PartialFunction[LayerExpr[E], ZLayer.Debug] = {
       case '{ ZLayer.Debug.tree }    => ZLayer.Debug.Tree
       case '{ ZLayer.Debug.mermaid } => ZLayer.Debug.Mermaid
     }
 
     '{
-      val trace: Trace = Tracer.newTrace
+      given Trace = Tracer.newTrace
 
       ${
         def typeToNode(tpe: TypeRepr): Node[TypeRepr, LayerExpr[E]] =
-          Node(Nil, List(tpe), tpe.asType match { case '[t] => '{ ZLayer.environment[t](trace) } })
+          Node(Nil, List(tpe), tpe.asType match { case '[t] => '{ ZLayer.environment[t] } })
 
         def composeH(lhs: LayerExpr[E], rhs: LayerExpr[E]): LayerExpr[E] =
           lhs match {
             case '{ $lhs: ZLayer[i, E, o] } =>
               rhs match {
                 case '{ $rhs: ZLayer[i2, E, o2] } =>
-                  val tag = Expr
-                    .summon[EnvironmentTag[o2]]
-                    .getOrElse(
-                      report.errorAndAbort(s"Cannot find EnvironmentTag[${TypeRepr.of[o2].show}] in implicit scope")
-                    )
-                  '{ $lhs.++($rhs)($tag) }
+                  rhs.asTerm match {
+                    case _: Ident => '{ $lhs.and($rhs)(summonInline) }
+                    case _        => '{ $lhs +!+ $rhs }
+                  }
```

Comment:
I'm not too familiar with `Ident`. How exactly does it differentiate between user defined code from macro-generated one?

Also, can we use `++` instead of `and` for consistency?

Reply:
`Ident` is short for "identifier" - that's a reference to a value as opposed to a more complex expression (composed layer in this case).